Document v4 of Messaging Bridge (#6831)
Particular/docs.particular.net: nservicebus/upgrades/bridge-3to4.md

---
title: Messaging Bridge Upgrade Version 3 to 4
component: Bridge
reviewed: 2024-09-12
isUpgradeGuide: true
upgradeGuideCoreVersions:
 - 8
 - 9
---

## ReplyToAddress header translation for failed messages

> Version 4 of the Messaging Bridge will translate the `NServiceBus.ReplyToAddress` message header for failed messages.

Reviewer: ```suggestion
Version 4 of the Messaging Bridge will by default translate the `NServiceBus.ReplyToAddress` message header for failed messages.
```

> This is a change from previous versions of the bridge, which defaulted to not translating this header.

Reviewer: ```suggestion
This is a change from previous bridge versions, in which the translation of the header had to be explicitly enabled.
```

> [!NOTE]

> > The `TranslateReplyToAddressForFailedMessages()` method is now obsolete and will be removed in the next major version of the bridge.  Calls to this method should be removed

Reviewer: ```suggestion
> The `TranslateReplyToAddressForFailedMessages()` method is now obsolete and will be removed in the next major version of the bridge. It is recommended to remove invocations of the method after upgrading to version 4.
```

> Since any endpoint can be the source of a failed message, the consequence of this change in behavior is that the all endpoints need to be registered with the bridge so that the `NServiceBus.ReplyToAddress` header can be translated.

Reviewer: ```suggestion
Since the header may contain a physical address of any endpoint in the system, the change in the behavior requires all endpoints in the system to be registered with the bridge. This ensures that the information necessary for the `NServiceBus.ReplyToAddress` header translation are available when messages are transferred by the bridge.
```

> [!WARNING]

> > If the `NServiceBus.ReplyToAddress` header cannot be translated by the bridge because the endpoint it refers to is not registered, then the failed messages

Reviewer: ```suggestion
> If the `NServiceBus.ReplyToAddress` header cannot be translated by the bridge (because the endpoint it refers to is not registered), then a message
```

>  will be moved to the `bridge.error` queue instead of the `error` queue that feeds in to ServiceControl.  Any messages in the `bridge.error` queue cannot be easily retried at this time,

Reviewer: ```suggestion
 will be moved to [the bridge error queue](/nservicebus/bridge/configuration.md#recoverability-error-queue). To prevent the need for retrying messages from the bridge error queue (which requires, a custom, transport-specific solution) it is recommended to register all endpoints in the system **before** upgrading to version 4.
```

>  so it is best to avoid this situation.

Reviewer: ```suggestion
```

> If it is problematic to register all endpoints with the bridge, the behavior can be disabled.

Reviewer: ```suggestion
Users who still want to prevent the need to register all endpoints with the bridge can use a dedicated API option:
```
